In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 8.9 MB 3.8 MB/s 
     |████████████████████████████████| 49 kB 5.1 MB/s 


$Exercise\;1$ 
\begin{align}
\min \ c_1 x_1 + c_2 x_2 + \ldots + c_N x_N, \\
{\rm{s.t.}} \; x_1 + x_2 + \ldots + x_N & \leq b_1, \\
w_1 x_1 + w_2 x_2 + \ldots + w_N x_N  & = b_2, \\
x_i & \leq u_i, \quad i=1,\ldots,N, \\
x_i & \geq l_i, \quad i=1,\ldots,N. \\
\end{align}
---



$b_1 = 161.2,$ \\
$b_2 = 54.6,$ \\
$c = [−8.1, 10.15, 30.5, 50.05, 0.05, 80.5, −0.25, −31.02, 50.65, 0.725, −0.8, 100.6];$ \\
$w = [0, 1.6, −2.02, 0, 1.01, −2.005, 0, 1.39, 1, −1.214, 0, −5.32];$ \\
$l = [−∞, 1, 0, 0, 7, 0, −∞, −∞, 1, 1, −5, 2],$ \\
$u = [4, 3, +∞, 2, 10, +∞, 13, 20, +∞, 21, 5, 60]$

$1.\;Solving\;the\;given\;problem:$

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
model = ConcreteModel()

In [ ]:
N = 12  

In [ ]:
obj_coef_c = np.array([-8.1, 10.15, 30.5, 50.05, 0.05, 80.5, -0.25, -31.02, 50.65, 0.725, -0.8, 100.6])

In [ ]:
constr1_coef_p = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
constr2_coef_w = np.array([0, 1.6, -2.02, 0, 1.01, -2.005, 0, 1.39, 1, -1.214, 0, -5.32])

constr1_rhs_b1 = 161.2
constr2_rhs_b2 = 54.6

In [ ]:
lower_bound = np.array([-np.inf, 1, 0, 0, 7, 0, -np.inf, -np.inf, 1, 1, -5, 2])
upper_bound = np.array([4, 3, np.inf, 2, 10, np.inf, 13, 20, np.inf, 21, 5, 60])

In [ ]:
col_indices = np.arange(N)
print(col_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11]


In [ ]:
model.x = Var(col_indices)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
model.constraints.add(sum(constr1_coef_p[j]*model.x[j] for j in col_indices) <= constr1_rhs_b1)

In [ ]:
#next add the constraint w1 x1 + w2 x2 + ...+ wN xN >= b2 
model.constraints.add(sum(constr2_coef_w[j]*model.x[j] for j in col_indices) == constr2_rhs_b2)

In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

1 Var Declarations
    x : Size=12, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :  -inf :  None :   4.0 : False :  True :  Reals
          1 :   1.0 :  None :   3.0 : False :  True :  Reals
          2 :   0.0 :  None :   inf : False :  True :  Reals
          3 :   0.0 :  None :   2.0 : False :  True :  Reals
          4 :   7.0 :  None :  10.0 : False :  True :  Reals
          5 :   0.0 :  None :   inf : False :  True :  Reals
          6 :  -inf :  None :  13.0 : False :  True :  Reals
          7 :  -inf :  None :  20.0 : False :  True :  Reals
          8 :   1.0 :  None :   inf : False :  True :

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160837 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 775.965100000001
  Upper bound: 775.965100000001
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 13
  Number of nonzeros: 21
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.011092901229858398
# ----------------------------------------------------------
#   Solution Information
# ------------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  775.9651000000001

Decision Variables
x[ 0 ] =  4.0
x[ 1 ] =  3.0
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  10.0
x[ 5 ] =  0.0
x[ 6 ] =  13.0
x[ 7 ] =  20.0
x[ 8 ] =  23.754
x[ 9 ] =  1.0
x[ 10 ] =  5.0
x[ 11 ] =  2.0

Constraints
constraints : Size=2
    Key : Lower : Body               : Upper
      1 :  None :             81.754 : 161.2
      2 :  54.6 : 54.599999999999994 :  54.6


$2.\; The\;optimal\;solution\;is: $ \\
$$x_1 =  4.0,\;x_2 =  3.0,\;x_3 =  0.0,\;x_4 =  0.0,\;x_5 =  10.0,\;x_6 =  0.0,\;x_7 =  13.0,\;x_8 =  20.0,\;x_9 =  23.754,\;x_10 =  1.0,\;x_{11} =  5.0,\; x_{12} =  2.0$$
$The\;value\;of\;the\;objective\;function\;at\;optimal\;solution\;is:\;775.9651000000001$